src/plugins.ts

```diff
@@ -101,43 +128,65 @@ export default class Plugins {
   }
 
   // if yarn.lock exists, fetch locked dependencies
-  async refresh(root: string, {prod = true}: {prod?: boolean} = {}): Promise<void> {
+  async refresh(
+    root: string,
+    {prod = true}: { prod?: boolean } = {},
+  ): Promise<void> {
     if (fs.existsSync(path.join(root, 'yarn.lock'))) {
       // use yarn.lock to fetch dependencies
-      await this.yarn.exec(prod ? ['--prod'] : [], {cwd: root, verbose: this.verbose})
+      await this.yarn.exec(prod ? ['--prod'] : [], {
+        cwd: root,
+        verbose: this.verbose,
+      })
     }
   }
 
   async link(p: string): Promise<void> {
     const c = await Config.load(path.resolve(p))
     CliUx.ux.action.start(`${this.config.name}: linking plugin ${c.name}`)
-    if (!c.valid && !this.config.plugins.find(p => p.name === '@oclif/plugin-legacy')) {
+    if (
+      !c.valid &&
+      !this.config.plugins.find(p => p.name === '@oclif/plugin-legacy')
+    ) {
       throw new Errors.CLIError('plugin is not a valid oclif plugin')
     }
 
+    // refresh will cause yarn.lock to install dependencies, including devDeps
     await this.refresh(c.root, {prod: false})
     await this.add({type: 'link', name: c.name, root: c.root})
+    // if the plugin has typescript in devDeps, we will also compile it
+    if ((await this.pjson()).devDependencies?.typescript) {
```

Comment:
`this.pjson()` doesn't get the `package.json` of the plugin being linked. It gets the `package.json` in the `oclif.config.dataDir`.
```suggestion
    if (c.pjson.devDependencies?.typescript) {
```